### PR TITLE
Add a user friendly error for merge conflicts on mix.lock.

### PR DESCRIPTION
**Pull request by tuvistavie**
As [proposed in the ML](https://groups.google.com/forum/#!topic/elixir-lang-core/ORl4QArbaxI), I added a user friendly message when `mix.lock` contains conflicts.

### CODE REVIEW

File: lib/mix/lib/mix/dep/lock.ex [hunk 1]
```diff
@@ -62,4 +63,11 @@ defmodule Mix.Dep.Lock do
   defp lockfile do
     Mix.Project.config[:lockfile]
   end
+
+  defp check_lockfile_conflicts(lockfile, info) do
+    if Regex.run(~r/<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|/, info) do
+      raise Mix.raise ~s(Your #{lockfile} contains merge conflicts.
```
Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: Just `Mix.raise/1`, not `raise` and `Mix.raise` :)

File: lib/mix/lib/mix/dep/lock.ex [hunk 2]
```diff
@@ -62,4 +63,11 @@ defmodule Mix.Dep.Lock do
   defp lockfile do
     Mix.Project.config[:lockfile]
   end
+
+  defp check_lockfile_conflicts(lockfile, info) do
+    if Regex.run(~r/<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|/, info) do
```
Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: Maybe we can use `Regex.match?/2` here?

File: lib/mix/test/mix/dep/lock_test.exs [hunk 3]
```diff
@@ -25,4 +25,21 @@ defmodule Mix.Dep.LockTest do
...
+    in_tmp context.test, fn ->
+      info = ~S"""
+        <<<<<<< HEAD
+        %{"foo": {:hex, :foo, "0.1.0"},
+        =======
+        %{"bar": {:hex, :bar, "0.1.0"},
+        >>>>>>> foobar
+        %{"baz": {:hex, :baz, "0.1.0"}}
+        """
+      File.write("mix.lock", info)
```
Comment (whatyouhide) [inline lib/mix/test/mix/dep/lock_test.exs]: Maybe we can shorten this to something like

```elixir
File.write "mix.lock", ~S"""
...
"""
```

? Wdyt?

File: lib/mix/lib/mix/dep/lock.ex (hunk 2 above)
Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: I think we should match on the beginning of the line.

Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: I think we can even just do a bunch of `String.contains?` calls tbh, a regex feels like overhead here :\

File: lib/mix/lib/mix/dep/lock.ex [hunk 4]
```diff
@@ -34,6 +34,7 @@ defmodule Mix.Dep.Lock do
   def read() do
     case File.read(lockfile) do
       {:ok, info} ->
+        check_lockfile_conflicts(lockfile, info)
```
Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: What about the name `assert_no_merge_conflicts_in_lockfile` as the name for this function?

File: lib/mix/lib/mix/dep/lock.ex (hunk 2 above)
Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: If we match on the beginning of the line, a regex seems simpler, if we don't, `String.contains?` should indeed be just fine. What do you think?

Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: I'd go with regex to be closer to the real conflicts output.

**Comment by tuvistavie** [comment]
@lexmag @whatyouhide Thank you very much for the feedback. I just updated the PR.

File: lib/mix/lib/mix/dep/lock.ex [hunk 5]
```diff
@@ -62,4 +63,13 @@ defmodule Mix.Dep.Lock do
   defp lockfile do
     Mix.Project.config[:lockfile]
   end
+
+  defp assert_no_merge_conflicts_in_lockfile(lockfile, info) do
+    if Regex.match?(~r/^<<<<<<<|^=======|^>>>>>>>|^\|\|\|\|\|\|\|/, info) do
```
Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: Let's use `^` once with `~r/^(?:<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|)/ ` pattern.

Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: Is there any info about `|||||||` marker? Never seen it before. :bowtie:

Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: Updated! :+1:

File: lib/mix/lib/mix/dep/lock.ex [hunk 6]
```diff
@@ -62,4 +63,13 @@ defmodule Mix.Dep.Lock do
   defp lockfile do
     Mix.Project.config[:lockfile]
   end
+
+  defp assert_no_merge_conflicts_in_lockfile(lockfile, info) do
+    if Regex.match?(~r/^(?:<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|)/, info) do
+      Mix.raise """
+      Your #{lockfile} contains merge conflicts.
+               Run `git checkout HEAD -- #{lockfile}` first to get a clean lock.
```
Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: I think this indentation needs to be fixed.

File: lib/mix/test/mix/dep/lock_test.exs [hunk 7]
```diff
@@ -25,4 +25,20 @@ defmodule Mix.Dep.LockTest do
...
+  test "raises a proper error for merge conflicts", context do
+    in_tmp context.test, fn ->
+      File.write "mix.lock", ~S"""
+        <<<<<<< HEAD
+        %{"foo": {:hex, :foo, "0.1.0"},
+        =======
+        %{"bar": {:hex, :bar, "0.1.0"},
+        >>>>>>> foobar
+        %{"baz": {:hex, :baz, "0.1.0"}}
+        """
```
Comment (whatyouhide) [inline lib/mix/test/mix/dep/lock_test.exs]: Contents of the heredoc should be indented on the same level as where the `"""` starts (so here, it should be brought two spaces left).

File: lib/mix/test/mix/dep/lock_test.exs [hunk 8]
```diff
@@ -25,4 +25,20 @@ defmodule Mix.Dep.LockTest do
...
+      assert_raise Mix.Error, ~r(Your mix.lock contains merge conflicts), fn ->
```
Comment (whatyouhide) [inline lib/mix/test/mix/dep/lock_test.exs]: I think the convention is to use `~r//` when there's no need to escape `/`s. Also, `mix\.lock` :)

File: lib/mix/lib/mix/dep/lock.ex (hunk 6 above)
Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: I was wondering what to do here, `** (Mix) ` is 9 chars, so I indented the same amount to get an output like this:

```sh
** (Mix) Your mix.lock contains merge conflicts.
         Run `git checkout HEAD -- mix.lock` first to get a clean lock.
```

What do you think?

Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: I don't think that's necessary; most of Elixir's source either uses `"""` and doesn't care about output like

```
** (Mix) Your mix.lock contains merge conflicts.
Run `git checkout HEAD -- mix.lock` first to get a clean lock.
```

(which is what I prefer personally), or uses no newlines at all and lets the host terminal do the splitting :).

File: lib/mix/test/mix/dep/lock_test.exs (hunk 8 above)
Comment (tuvistavie) [inline lib/mix/test/mix/dep/lock_test.exs]: Did not know about this convention, thank you!
Completely forgot to escape the dot :sweat:

File: lib/mix/lib/mix/dep/lock.ex (hunk 6 above)
Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: Ok, I'll go with the newline without indentation then, thank you!

**Comment by whatyouhide** [comment]
(very very small nitpick, sorry, but no trailing period in the commit message :bowtie:)

**Comment by tuvistavie** [comment]
@whatyouhide Oh, I think I had a trailing period in all my other commit messages here :scream: I'll avoid them from now on, thank you!

**Comment by whatyouhide** [comment]
This looks good to me, great job @tuvistavie :)

File: lib/mix/lib/mix/dep/lock.ex [hunk 9]
```diff
@@ -62,4 +63,13 @@ defmodule Mix.Dep.Lock do
   defp lockfile do
     Mix.Project.config[:lockfile]
   end
+
+  defp assert_no_merge_conflicts_in_lockfile(lockfile, info) do
+    if Regex.match?(~r/^(?:<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|)/, info) do
```
Comment (josevalim) [inline lib/mix/lib/mix/dep/lock.ex]: What about `info =~ ~r/^(?:<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|)/`?

File: lib/mix/lib/mix/dep/lock.ex [hunk 10]
```diff
@@ -62,4 +63,13 @@ defmodule Mix.Dep.Lock do
...
+      Mix.raise """
+      Your #{lockfile} contains merge conflicts.
+      Run `git checkout HEAD -- #{lockfile}` first to get a clean lock.
```
Comment (josevalim) [inline lib/mix/lib/mix/dep/lock.ex]: Let's not use `"""` because it adds an uncessary new line at the end. What do you think about this error messages:

```elixir
Mix.raise "Your #{lockfile} contains merge conflicts. If you are using git, " <>
          "run `git checkout HEAD -- #{lockfile}` to revert local changes"
```

?

**Comment by tuvistavie** [comment]
@josevalim Thank you for the feedback, I just updated the PR.

File: lib/mix/lib/mix/dep/lock.ex [hunk 11]
```diff
@@ -62,4 +63,11 @@ defmodule Mix.Dep.Lock do
   defp lockfile do
     Mix.Project.config[:lockfile]
   end
+
+  defp assert_no_merge_conflicts_in_lockfile(lockfile, info) do
+    if info =~ ~r/^(?:<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|)/ do
```
Comment (josevalim) [inline lib/mix/lib/mix/dep/lock.ex]: Actually, what about: `String.contains?(info, ~w(<<<<<<< ======= >>>>>>> |||||||))`. Cleaner and very likely faster.

Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: This is what we discussed here: https://github.com/elixir-lang/elixir/pull/4482#discussion_r58687657
With `String.contains?` it is harder to match on the beginning of the line.
What do you think?

Comment (ericmj) [inline lib/mix/lib/mix/dep/lock.ex]: `match?({0, _}, :binary.match(info, ["<<<", "===", ">>>", ...]))` will match only on the front. You can also do normal binary pattern matching:

```elixir
defp match?("<<<" <> _), do: true
defp match?("===" <> _), do: true
defp match?(_), do: false
```

It's a good idea to avoid regular expressions when possible.

File: lib/mix/lib/mix/dep/lock.ex [hunk 12]
```diff
@@ -62,4 +63,11 @@ defmodule Mix.Dep.Lock do
...
+      Mix.raise "Your #{lockfile} contains merge conflicts. If you are using git, " <>
+                "run `git checkout HEAD -- #{lockfile}` to revert local changes"
```
Comment (ericmj) [inline lib/mix/lib/mix/dep/lock.ex]: I don't like recommending `git checkout` since that will make you lose the changes.

File: lib/mix/lib/mix/dep/lock.ex (hunk 11 above)
Comment (ericmj) [inline lib/mix/lib/mix/dep/lock.ex]: Shouldn't it be enough to only match on one of these?

Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: I think it should be enough, but bundler was using this regex, so I am not sure if there is some edge case with mercurial, bazaar or some other VCS where there could be only one of these.

> `match?({0, _}, :binary.match(info, ["<<<", "===", ">>>", ...]))` will match only on the front. You can also do normal binary pattern matching:

Thank you, you just made me realize there was a bug in the current code, the regex should be `~r/^(?:<<<<<<<|=======|>>>>>>>|\|\|\|\|\|\|\|)/m`, without the `m` modifier the `^` does not behave as intended.
I think your proposals would work without the `m` modifier, but with it, we would need to split the string right?

File: lib/mix/lib/mix/dep/lock.ex (hunk 12 above)
Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: I agree on this. Maybe we can just say "fix the merge conflicts" or something along these lines.

Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: Do you think we should rather only notify about the conflict without suggesting to `git checkout`?

File: lib/mix/lib/mix/dep/lock.ex (hunk 11 above)
Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]: I think we can safely start with supporting just git and think about other VCs if people ask for them.

File: lib/mix/lib/mix/dep/lock.ex (hunk 12 above)
Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: What about something like

```
Mix.raise "Your #{lockfile} contains merge conflicts. Please resolve the conflicts " <>
          "and run the command again."
```

File: lib/mix/lib/mix/dep/lock.ex (hunk 11 above)
Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: Ok, maybe we can go with `~r/^<<<<<<</m` then? Git conflicts should all have this marker.

Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: We can drop `\|\|\|\|\|\|\|` marker, it's optional and always present together with other markers, not on its own.

Comment (josevalim) [inline lib/mix/lib/mix/dep/lock.ex]: Well, we don't need to match it at the beginning of the line. We don't expect <<<<... or ====... to even in the middle of lines, so String.contains? is fine for the performance and readability wins.

Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: But people seems like ASCII art _(just kidding)_.

Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: I changed to `String.contains?` and remove the pipe marker.

**Comment by whatyouhide** [comment]
I'm in favour of removing the `git checkout` suggestion from the error message as well; I think just pointing the user to the merge conflict may already be enough?

File: lib/mix/lib/mix/dep/lock.ex (hunk 12 above)
Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: > Please resolve the conflicts and run the command again

Sounds good :+1:.

**Comment by josevalim** [comment]
Ok, let's change the message and ship it. Thank you @tuvistavie! 
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by tuvistavie** [comment]
I just updated the message!

File: lib/mix/lib/mix/dep/lock.ex [hunk 13]
```diff
@@ -62,4 +63,11 @@ defmodule Mix.Dep.Lock do
   defp lockfile do
     Mix.Project.config[:lockfile]
   end
+
+  defp assert_no_merge_conflicts_in_lockfile(lockfile, info) do
+    if String.contains?(info, ~w(<<<<<<< ======= >>>>>>>)) do
+      Mix.raise "Your #{lockfile} contains merge conflicts. Please resolve the conflicts " <>
+                "and run the command again."
```
Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: We use no trailing period in error messages. :bowtie:

Comment (tuvistavie) [inline lib/mix/lib/mix/dep/lock.ex]: Oh, sorry for that :sweat:

Comment (lexmag) [inline lib/mix/lib/mix/dep/lock.ex]: No worries, I'll fix that. :yellow_heart:
